### ISSUE TITLE
CA1010 false positive

### ISSUE THREAD

**Issue by kentcb** [dotnet/roslyn-analyzers#1325]
#### Analyzer package

Microsoft.CodeAnalysis.FxCopAnalyzers

#### Analyzer

CA1010

#### Repro steps

```C#
// external code, out of my control
public class ThirdPartyType : IEnumerable
{
    ...
}

public class MyClass : ThirdPartyType
{
}
```

#### Expected behavior

No violation, since it is not my code violating the rule.

#### Actual behavior

```
error CA1010: Collections should implement generic interface
```

**Comment by mavasani** [comment]
This should be fixed in 15.5

**Comment by 333fred** [comment]
This is actually a break from FxCop behavior, but we agree that this should be changed.

**Comment by sharwell** [comment]
I'm not sure I agree with the change. For example, would this change cause a warning to no longer be reported on a type that extends [`ArrayList`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.arraylist)?

**Comment by 333fred** [comment]
@sharwell we can probably update the change to flag anything that directly inherits from a type in `System.Collections`. However, I'm still of the opinion that we shouldn't be flagging all descendants of these things, as that's a good way to create a lot of noise, especially in scenarios like the original one here that are out of a user's control.

**Comment by sharwell** [comment]
@333fred I wouldn't say it's out of the user's control. It's pretty easy to add an `IEnumerable<T>` implementation if you already have an `IEnumerable` implementation on the object.

**Comment by 333fred** [comment]
I would argue that in this scenario, there's no point to them adding that implementation. Likely the user is creating this object to extend an existing framework, which won't take advantage of the `IEnumerable<T>` implementation, and they quite possibly won't use it themselves. Additionally, this change ensures that this warning doesn't proliferate through a type hierarchy that a user has defined, reducing noise to where the user actually needs to see it.

**Comment by sharwell** [comment]
@333fred the exception appears to be a deviation from the primary intent of the rule. It's alerting the user that an implementation appears to be out-of-date with respect to current practices, and to verify that the code is correct. In the original example, updating `MyClass` to extend `UpdatedThirdPartyType` instead of a legacy type `ThirdPartyType` might fix the issue. For the rare case where the user has no option, suppressions can be used or the rule simply disabled.

**Comment by 333fred** [comment]
@sharwell we agree on what the intent of the rule, but it feels to me like the noise-signal ratio will be much improved with this change. @mavasani, did you get any metrics on scenarios like this when you were running code on popular repositories? Is this an issue that will affect more than a couple of users?

**Comment by mavasani** [comment]
Newtonsoft did have couple of reports for types inheriting from ArrayList. I hope to execute on other repos soon.

**Comment by mavasani** [comment]
I think we should either implement Fred’s suggestion of reducing the noise OR give a different error message for an indirect violation, something that recommends inheriting from a variant of the type if possible or suppress if required. I just got an internal customer raise a similar request: they are deriving from JObject and get this report and are not sure if it is okay to suppress or not.

**Comment by 333fred** [comment]
@sharwell, any more comments here? We'd like to do something to reduce the noise.

**Comment by mavasani** [comment]
@333fred Can we have different error messages for 2.6.0 release and discuss if we need to avoid flagging this error for such cases in 2.7.0?

**Comment by 333fred** [comment]
Yeah, that shouldn't be too difficult.

**Comment by 333fred** [comment]
@mavasani @sharwell it occurs to me, now that this warning isn't reported for internal types and is only reported on public types, that perhaps this distinction isn't necessary. If are publicly exposing a type, you should probably implement these interfaces, regardless of where the warning is coming from. Thoughts? We should definitely improve the messaging here (we should say what interface you should implement) regardless.

**Comment by sharwell** [comment]
@333fred I am in favor of not changing the logic at this time. The messaging could be changed if it will help users.
